### ISSUE TITLE
Email -> Magic Link Login 🔗 

### ISSUE THREAD

**Issue by nelsonic** [dwyl/auth#223]
As noted by @ndrean in https://github.com/dwyl/learn-postgresql/issues/84#issuecomment-1277916838 💬 
We don't currently have the concept of Magic Links in this project. 🪄 
My reasoning was that we don't _want_ people clicking links in email because it promotes [**`Phising`**](https://en.wikipedia.org/wiki/Phishing) ... 🙄 
But we ***`do`*** already have email verification. 
e.g: [templates/verify.txt#L6-L7](https://github.com/dwyl/aws-ses-lambda/blob/33f1796ec2df36095a21784056638e80b6a2ca5b/templates/verify.txt#L6-L7)

# Todo

+ [ ] Gage the demand from other people in the community for this feature before adding it. Might be that we only consider it as part of the re-build with `gen.auth` #133 -> #207 

P.S: Just tried the [`email`](https://github.com/dwyl/email) send function and it still all works. 🤯 
![image](https://user-images.githubusercontent.com/194400/195687912-f5a5fa14-81d3-47b3-b4ec-8f398c2098aa.png)

I probably shouldn't be surprised because nothing has changed and it worked fine last time I checked ... ⏳ 💭 😜

**Comment by iteles** [comment]
As a technical tool, I love magic links. They're very cool.

As a _layman user_ of apps I personally am not a fan of magic links as I have seen them historically implemented from UX perspective. I used to use them back in my slack days (4 years ago) and they were cool the first time, annoying the rest.

I'm annoyed by having to go back to my email when trying to log into an app and especially so when I have to request magic links for multiple different devices (usually they expire after one click) and having to do this process continually across all devices when the login expires after a few months is frustrating.

I already have password management in place (and these days, most people do in their browsers anyway) and this messes with that system. 

I also can never remember which email address I signed up to something with but this could still be managed by the password manager so is less of an issue.

Interested to see if I'm in the minority and magic links are all the rage!

**Comment by nelsonic** [comment]
@iteles I feel that Magic Links are a **UX-_bandaid_** 🩹  more than a long-term solution to the auth challenge. 💭 
We've chosen to focus on **OAuth** Providers [Google and GitHub] for now as the _easiest/fastest_ way to authenticate. 🚀 
I think that if we add **`Microsoft` Auth** https://github.com/dwyl/elixir-auth-microsoft/issues/1 we will have _all_ the bases covered from a [corporate] **`OAuth`** perspective ... :shipit: 
95% of people have a Google or Microsoft Account. 
Then if we add Sign in with Apple #37 🍏 we will reach 99% of people with mobile devices. 💯 
My hunch is that _very_ few people will opt to use an email address to authenticate. 📉 
And those people might not be our "target customers" for the **`App`** ... 🤷‍♂️ 

But this issue is open for feedback and if enough people [that will *pay* to use the **`App`**] want it, we can build it. 👌

**Comment by ndrean** [comment]
For mobile devices, I heard that 3/4 of people don't have an email address .... so depends on your target. Indeed, OAuth is the way to go I believe, but one giant is missing, Facebook. I looked at it the other day and did not completely understand nor had the patience to implement the [server-side process](https://developers.facebook.com/docs/facebook-login/guides/advanced/manual-flow)🙄

**Comment by nelsonic** [comment]
We are keen on helping people **migrate _off_** [**`Facebook`**](https://github.com/dwyl/learn-react/issues/23#issuecomment-647521076) (eventually) once our **`App`** has a few more features.
But we haven't even listed them on the [**roadmap**](https://github.com/dwyl/product-roadmap) yet, because we're avoiding _distracting_ ourselves with "social" features ...
If you want to take a look at this: https://github.com/dwyl/elixir-auth-facebook/issues/21 👀 

Otherwise let's keep the discussion of this thread specific to Magic Links. 🔗 🪄 👌

**Comment by ndrean** [comment]
You are wise🤔.  I believe passwordless is straightforward: a signed token based on a user_id (new_or_create) which is saved with "pending" status and on reception to an endpoint, the query string contains a token and is verified for the signature and decoded and a lookup at the user_id to check for the token, and if ok, generate a new one, save it and pass status "confirmed". I made something like this for my little app.

**Comment by ndrean** [comment]
Do you admit `links` as function components?

**Comment by nelsonic** [comment]
Was recently reminded of the "_right_" way of doing `email` verification 
`while` I was setting up `Plausible/Analytics`: https://github.com/dwyl/atm/issues/29#issuecomment-1418767844
![image](https://user-images.githubusercontent.com/194400/218268082-188c9b01-f6ac-4e1b-8816-6194b356c08f.png)

![image](https://user-images.githubusercontent.com/194400/218268089-420f7fbc-4ab4-401c-a861-9760fdff1cd7.png)

![image](https://user-images.githubusercontent.com/194400/218268132-24c0a1dd-3584-4ede-a361-89df4619c38a.png)

This is the approach we will be taking. 
Yes, there is an extra step; copy-pasting or manually typing the verification code. 
It not only **_feels_ more secure** to the person verifying their account, it **_is_ more secure**.
This is one of the _few_ instances where the UX is _improved_ by having an extra step.  

"_**Never** click links in email_" ~ https://www.kaspersky.com/resource-center/preemptive-safety/phishing-prevention-tips

**Comment by ndrean** [comment]
You may be right about these evil links.🙄 At this moment, I have plenty of fake emails with links to follow coming from fake accounts such as DHL, UPS etc...

Sending a code via email makes the code slimmer compared to the magic link; you don't have the extra endpoint with the URL parsing step. You only need to insert a tuple  "email, code_sent" into the DB and store the active email in the session, don't you? How does this work for mobile where you don't have a session?

**Comment by ndrean** [comment]
Probably something to watch: <https://hexdocs.pm/webauthn_components/readme.html>

**Comment by nelsonic** [comment]
Definitely. One to watch. https://webauthn.guide

**Comment by ndrean** [comment]
[<img width="879" alt="Screenshot 2023-02-15 at 12 51 44" src="https://user-images.githubusercontent.com/6793008/219020179-c32a8795-e76d-4402-8a84-46ac87e9839b.png">](https://www.youtube.com/watch?v=krbriAecqUk&t=41s)

**Comment by nelsonic** [comment]
Yeah, watched that a while back. still very much "beta" though ... 
I'm very much of a "Wait and see if it becomes mainstream" when it comes to auth systems now.
Glad someone `else` is doing work to innovate in the space. 
But ultimately use `OAuth` 90%+ of the time.
